src/main/java/me/botsko/prism/database/sql/SQLPrismDataSource.java

```diff
@@ -207,81 +224,57 @@ public void setupDatabase(ActionRegistry actionRegistry) {
-            if (st != null)
-                try {
-                    st.close();
-                } catch (final SQLException ignored) {
-                }
-            if (conn != null)
-                try {
-                    conn.close();
-                } catch (final SQLException ignored) {
-                }
         }
     }
 
+    /**
+     * Add action to db.
+     * @param actionName String
+     */
     public void addActionName(String actionName) {
 
-        if (Prism.prismActions.containsKey(actionName))
+        if (Prism.prismActions.containsKey(actionName)) {
             return;
-
-        Connection conn = null;
-        PreparedStatement s = null;
-        ResultSet rs = null;
-        try {
-
-            conn = database.getConnection();
-            s = conn.prepareStatement("INSERT INTO " + getPrefix() + "actions (action) VALUES (?)",
-                    Statement.RETURN_GENERATED_KEYS);
+        }
+        try (
+                Connection conn = database.getConnection();
+                PreparedStatement s = conn.prepareStatement("INSERT INTO " + prefix + "actions (action) VALUES (?)",
+                        Statement.RETURN_GENERATED_KEYS)
+        ) {
             s.setString(1, actionName);
             s.executeUpdate();
-
-            rs = s.getGeneratedKeys();
+            ResultSet rs = s.getGeneratedKeys();
             if (rs.next()) {
                 log.info("Registering new action type to the database/cache: " + actionName + " " + rs.getInt(1));
                 Prism.prismActions.put(actionName, rs.getInt(1));
             } else {
                 throw new SQLException("Insert statement failed - no generated key obtained.");
             }
-        } catch (final SQLException ignored) {
+            rs.close();
+        } catch (final SQLException e) {
+            handleDataSourceException(e);
 
-        } finally {
-            if (rs != null)
-                try {
-                    rs.close();
-                } catch (final SQLException ignored) {
-                }
-            if (s != null)
-                try {
-                    s.close();
-                } catch (final SQLException ignored) {
-                }
-            if (conn != null)
-                try {
-                    conn.close();
-                } catch (final SQLException ignored) {
-                }
         }
     }
 
     protected void cacheActionPrimaryKeys() {
-        Connection conn = null;
-        PreparedStatement s = null;
-        ResultSet rs = null;
-        try {
-            conn = getConnection();
-            s = conn.prepareStatement("SELECT action_id, action FROM " + getPrefix() + "actions");
-            rs = s.executeQuery();
 
+        try (
+                Connection conn = getConnection();
+                PreparedStatement s = conn.prepareStatement(
+                        "SELECT action_id, action FROM " + prefix + "actions");
+                ResultSet rs = s.executeQuery()
```

Comment:
Does `ResultSet rs = s.executeQuery()` need to end with a semicolon?  If it does, compilation should fail ... and I'm sure you're compiled before committing, :wink:

Reply:
Oh !  I see; this is like a `using` statement in C#, but instead a localized try/catch that exists during the scope of `rs`